spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarTemplate.java

```diff
@@ -227,12 +245,15 @@ private Observation newObservation(PulsarMessageSenderContext senderContext) {
 				DefaultPulsarTemplateObservationConvention.INSTANCE, () -> senderContext, this.observationRegistry);
 	}
 
-	private Producer<T> prepareProducerForSend(@Nullable String topic, T message, @Nullable Schema<T> schema,
+	private Producer<T> prepareProducerForSend(@Nullable String topic, @Nullable T message, @Nullable Schema<T> schema,
 			@Nullable Collection<String> encryptionKeys, @Nullable ProducerBuilderCustomizer<T> producerCustomizer)
 			throws PulsarClientException {
-		if (schema == null) {
-			schema = Objects.requireNonNull(this.schemaResolver.getSchema(message),
-					"Schema must not be null - expecting at least a default schema");
+		Schema<T> resolvedSchema = schema == null ? this.schemaResolver.getSchema(message) : schema;
+		if (resolvedSchema == null) {
+			if (this.schemaResolver.getClass().equals(DefaultSchemaResolver.class)) {
```

Comment:
Ah yeh, this again. 😆 

Should we add a `SchemaResolver.exceptionForUnresolvableMesage(message)` and `TopicResolver.exceptionForUnresolvable(message)` ? Or something like that?

Reply:
**OR** we could change the TR signature to return ~ `ResolveTopicResult` which could contain an optional resolved topic OR a reason why it could not be resolved. That would remove the need for special message casing here and allow custom impls ability to tell users about why their custom resolver did not work. Wdyt?

Reply:
I also used `Resolved<T>` for the schema